src/keymgmt.c

```diff
@@ -1013,63 +1041,149 @@ static void *p11prov_rsapss_gen(void *genctx, OSSL_CALLBACK *cb_fn,
 {
     struct key_generator *ctx = (struct key_generator *)genctx;
     CK_BBOOL token_supports_allowed_mechs = CK_TRUE;
-    P11PROV_OBJ *key = NULL;
+    bool set_quirk = false;
+    void *key = NULL;
     CK_RV ret;
 
-    key = p11prov_rsa_gen(genctx, cb_fn, cb_arg);
-    if (!key) {
-        return NULL;
-    }
-
-    /* params could already be caying pss restriction. If allowed_types
-     * is already set, skip setting defaults */
-    if (!ctx->data.rsa.allowed_types) {
-        ret = set_default_rsapss_mechanisms(ctx);
-        if (ret != CKR_OK) {
-            P11PROV_raise(ctx->provctx, ret, "Failed to get pss params");
-            goto done;
-        }
-    }
-
+    /* check if we can add CKA_ALLOWED_MECHANISMS at all */
     ret = p11prov_token_sup_attr(ctx->provctx, p11prov_obj_get_slotid(key),
                                  GET_ATTR, CKA_ALLOWED_MECHANISMS,
                                  &token_supports_allowed_mechs);
     if (ret != CKR_OK) {
-        P11PROV_raise(ctx->provctx, ret, "Failed to probe quirk");
+        P11PROV_raise(ctx->provctx, ret,
+                      "Failed to probe CKA_ALLOWED_MECHANISMS quirk");
         goto done;
     }
 
-    if (token_supports_allowed_mechs == CK_TRUE) {
-        CK_ATTRIBUTE template[] = {
-            { CKA_ALLOWED_MECHANISMS, ctx->data.rsa.allowed_types,
-              ctx->data.rsa.allowed_types_size },
-        };
-        CK_ULONG tsize = 1;
+    if (token_supports_allowed_mechs) {
+        /* We always want to restrict PSS keys to just the PSS mechanisms.
+         * If a specific restriction is not already set then set defaults */
+        if (!ctx->data.rsa.allowed_types) {
```

Comment:
This is just a nit regarding the naming, the `allowed_types` might sounds a bit confusing. I would prefer `allowed_mechs` to be clear what it is on the first sight.

Reply:
Well we have "mechanism types" listed here, so "types" is just fine imo.
Not planning on changing that structure in this PR at any rate.